File: binance4j-blvt/src/main/java/com/binance4j/blvt/client/BLVTClient.java
package com.binance4j.blvt.client;

import com.binance4j.blvt.param.LimitInfoParams;
import com.binance4j.blvt.param.RedemptionParams;
import com.binance4j.blvt.param.SubscriptionParams;
import com.binance4j.blvt.param.TokenInfoParams;
import com.binance4j.blvt.param.TransactionRecordParams;
import com.binance4j.blvt.request.GetLimitInfoRequest;
import com.binance4j.blvt.request.GetRedemptionsRequest;
import com.binance4j.blvt.request.GetSubscriptionsRequest;
import com.binance4j.blvt.request.GetTokenInfoRequest;
import com.binance4j.blvt.request.RedeemRequest;
import com.binance4j.blvt.request.SubscribeRequest;
import com.binance4j.core.client.RestClient;
import com.binance4j.core.param.Params;
import com.binance4j.core.param.TimeFrame;

/**
 * Api client for the BLVT endpoints
 * 
 * @see <a href=
 *      "https://binance-docs.github.io/apidocs/spot/en/#blvt-endpoints">Documentation</a>
 */
public class BLVTClient extends RestClient<BLVTMapping> {
	/**
	 * @param key    API public key.
	 * @param secret API secret key.
	 */
	public BLVTClient(String key, String secret) {
		super(BLVTMapping.class, key, secret);
	}

	/**
	 * Get BLVT Info.
	 *
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetTokenInfoRequest getTokenInfo(TokenInfoParams params) {
		return new GetTokenInfoRequest(service.getTokenInfo(params.toMap()));
	}

	/**
	 * Get BLVT Info.
	 *
	 * @return The request to execute.
	 */
	public GetTokenInfoRequest getTokenInfo() {
		return new GetTokenInfoRequest(service.getTokenInfo(new TokenInfoParams(null).toMap()));
	}

	/**
	 * Subscribe BLVT.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public SubscribeRequest subscribe(SubscriptionParams params) {
		return new SubscribeRequest(service.subscribe(params.toMap()));
	}

	/**
	 * Get subscription record.
	 * 
	 * @param params    Request params.
	 * @param timeFrame Time search params.
	 * @return The request to execute.
	 */
	public GetSubscriptionsRequest getSubscriptions(TransactionRecordParams params, TimeFrame timeFrame) {
		return new GetSubscriptionsRequest(service.getSubscriptions(Params.merge(params, timeFrame)));
	}

	/**
	 * Get subscription record.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetSubscriptionsRequest getSubscriptions(TransactionRecordParams params) {
		return new GetSubscriptionsRequest(service.getSubscriptions(params.toMap()));
	}

	/**
	 * Get subscription record.
	 * 
	 * @return The request to execute.
	 */
	public GetSubscriptionsRequest getSubscriptions() {
		return new GetSubscriptionsRequest(service.getSubscriptions(new TransactionRecordParams(null, null).toMap()));
	}

	/**
	 * Redeem BLVT.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public RedeemRequest redeem(RedemptionParams params) {
		return new RedeemRequest(service.redeem(params.toMap()));
	}

	/**
	 * Get redemption record.
	 * 
	 * @param params    Request params.
	 * @param timeFrame Time search params.
	 * @return The request to execute.
	 */
	public GetRedemptionsRequest getRedemptions(TransactionRecordParams params, TimeFrame timeFrame) {
		return new GetRedemptionsRequest(service.getRedemptions(Params.merge(params, timeFrame)));
	}

	/**
	 * Get redemption record.
	 * 
	 * @param params Request params.
	 * @return The request to execute.
	 */
	public GetRedemptionsRequest getRedemptions(TransactionRecordParams params) {
		return new GetRedemptionsRequest(service.getRedemptions(params.toMap()));
	}

	/**
	 * Get redemption record.
	 * 
	 * @return The request to execute.
	 */
	public GetRedemptionsRequest getRedemptions() {
		return new GetRedemptionsRequest(service.getRedemptions(new TransactionRecordParams(null, null).toMap()));
	}

	/**
	 * Get user limit info.
	 * 
	 * @param params Request params.
	 * @return The request to LimitInfo
	 */
	public GetLimitInfoRequest getLimitInfo(LimitInfoParams params) {
		return new GetLimitInfoRequest(service.getLimitInfo(params.toMap()));
	}

	/**
	 * Get user limit info.
	 * 
	 * @return The request to LimitInfo
	 */
	public GetLimitInfoRequest getLimitInfo() {
		return new GetLimitInfoRequest(service.getLimitInfo(new LimitInfoParams(null).toMap()));
	}
}

File: binance4j-blvt/src/main/kotlin/com/binance4j/blvt/client/BLVTClient.kt
/*
 * MIT License
 *
 * Copyright (c) 2022 Binance4j
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NON INFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.binance4j.blvt.client

import com.binance4j.blvt.dto.*
import com.binance4j.blvt.param.*
import com.binance4j.core.Request
import com.binance4j.core.client.RestClient
import com.binance4j.core.param.Params
import com.binance4j.core.param.TimeFrame


/**
 * Api client for the BLVT endpoints.
 * @param key    API public key.
 * @param secret API secret key.
 *
 * [Binance doc](https://binance-docs.github.io/apidocs/spot/en/#blvt-endpoints)
 */
class BLVTClient(key: String, secret: String) : RestClient<BLVTMapping>(BLVTMapping::class.java, key, secret) {
    /**
     * Get BLVT Info.
     * @param params Request params.
     * @return The request to execute.
     */
    @JvmOverloads
    fun getTokenInfo(params: TokenInfoParams? = TokenInfoParams()): Request<List<Token>> = Request(service.getTokenInfo(Params.merge(params)))

    /**
     * Get subscription record.
     * @param params    Request params.
     * @param timeFrame Time search params.
     * @return The request to execute.
     */
    @JvmOverloads
    fun getSubscriptions(params: TransactionRecordParams? = TransactionRecordParams(), timeFrame: TimeFrame? = null): Request<List<Subscription>> =
        Request(service.getSubscriptions(Params.merge(params, timeFrame)))

    /**
     * Get redemption record.
     * @param params    Request params.
     * @param timeFrame Time search params.
     * @return The request to execute.
     */
    @JvmOverloads
    fun getRedemptions(params: TransactionRecordParams? = TransactionRecordParams(), timeFrame: TimeFrame? = null): Request<List<Redemption>> =
        Request(service.getRedemptions(Params.merge(params, timeFrame)))

    /**
     * Get user limit info.
     * @param params Request params.
     * @return The request to execute.
     */
    @JvmOverloads
    fun getLimitInfo(params: LimitInfoParams? = LimitInfoParams()): Request<List<LimitInfo>> = Request(service.getLimitInfo(Params.merge(params)))


    /**
     * Subscribe BLVT.
     * @param params Request params.
     * @return The request to execute.
     */
    fun subscribe(params: SubscriptionParams): Request<SubscriptionResponse> = Request(service.subscribe(params.toMap()))


    /**
     * Redeem BLVT.
     * @param params Request params.
     * @return The request to execute.
     */
    fun redeem(params: RedemptionParams): Request<RedemptionResponse> = Request(service.redeem(params.toMap()))
}
